posydon/popsyn/binarypopulation.py

```diff
@@ -317,7 +317,8 @@ def _safe_evolve(self, **kwargs):
                 binary = self.manager.generate(index=index, **self.kwargs)
             binary.properties = self.population_properties
 
-            with warnings.catch_warnings(record=True) as w:
+            with Catch_POSYDON_Warnings(record=True) as cpw:
```

Comment:
If the new Catch_POSYDON_Warnings() function still suppresses the warnings being written to standard error if record=True, then we need to explicitly print the warnings at the end of the code block, like I was doing before, for them to show-up consistently.

Reply:
It does, what the code tells: the filter is default (show the first warning of a type and suppress the later ones, it was a bug before, that it wasn't doing that always).
You can add
```
if self.kwargs.get("warnings_verbose", False):
    from posydon.utils.posydonwarning import AllPOSYDONWarnings
    AllPOSYDONWarnings()
```
(You can move the import up and merge it with the other if you like)

In the case of running multiple populations: it will work as you intend when you really run a new population (new python instance), if you run it again in a notebook, the records of the warning of the first run are still there, hence we talk about recurring warnings.

Removing the `reset_cache()` doesn't change anything on the filtering. So surly it can't effect any warnings, which are filtered out.

The new `Catch_POSYDON_Warnings()` does not print the warnings at the moment they occur, instead they will be printed to standard error at the end of the content (end of the with block). The statement `reset_cache()` suppresses that print out in case there are no warnings wanted.

Reply:
Ok, I guess I am confused on a couple points then. I tried adding-in the code block you suggested yesterday, and that did not fix the filtering. Ideally we if warnings_verbose=True, we want the first instance of warnings to show-up for every binary separately, which is what was happening before, but now those are being suppressed. 

Possible related to your second point, I think before, in the python version of catch_warnings, the warning registry was being cleared every time a population was run, even in a notebook without refreshing the kernel. So I am confused what changed where this is not happening anymore. 

For printing the warnings, nothing was being printed at the end of the content block when I was running populations even when I removed the reset_cache() block, which is why I am confused about the functionality.

Reply:
OK, there are two things (buggy), which have took effect here with the python warnings purely:

1. The warnings may include some binary specific detail, e.g. the index and so the python warnings call them based on text different.
2. There was a filter change made: python warnings clear their registry whenever a filter change is done, independent on the fact whether any warning in the registry would have been effected by the filter change.

The suggested code block should go at the beginning directly behind the with statement.
To debug things you can try
```
import warnings
print(warnings.filters)
```
to see whether the filter was successfully added.
To check the warnings registry you can use
```
from posydon.utils.posydonwarning import print_stats
print_stats()
```
To see was is recorded in the POSYDON warnings registry.

Reply:
I have added an option to use a context registry instead of the global one. Hence, it will issue the first warning in each context if used, without a need to change filters.